Тетрадка предназначена для студентов образвательного проекта Booster MDS использование вне проекта запрещено.

##### Задания не несут смысловой нагрузки для реальной жизни, они сделаны лишь для того, чтоб научить использовать Pandas

#### Каждое задание оценивает в 1 балл, для успешной сдачи задания нужно набрать 12 баллов

This database contains 76 attributes, but all published experiments refer to using a subset of 14 of them. In particular, the Cleveland database is the only one that has been used by ML researchers to this date. The "goal" field refers to the presence of heart disease in the patient. It is integer valued from 0 (no presence) to 1.

https://www.kaggle.com/ronitf/heart-disease-uci

___

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('heart_pandas.csv')

In [3]:
df.head(5)

age     sex   chest_pain_type  resting_blood_pressure  cholesterol  \
0   63    male  non-anginal pain                     145          233   
1   37    male   atypical angina                     130          250   
2   41  female    typical angina                     130          204   
3   56    male    typical angina                     120          236   
4   57  female    typical angina                     120          354   

     fasting_blood_sugar               rest_ecg  max_heart_rate_achieved  \
0  greater than 120mg/ml                 normal                      150   
1    lower than 120mg/ml  ST-T wave abnormality                      187   
2    lower than 120mg/ml                 normal                      172   
3    lower than 120mg/ml  ST-T wave abnormality                      178   
4    lower than 120mg/ml  ST-T wave abnormality                      163   

  exercise_induced_angina  st_depression   st_slope  num_major_vessels  \
0                      no            2.3  upsloping                  0   
1                      no            3.5  upsloping                  0   
2                      no            1.4       flat                  0   
3                      no            0.8       flat                  0   
4                     yes            0.6       flat                  0   

    thalassemia  target  
0        normal       1  
1  fixed defect       1  
2  fixed defect       1  
3  fixed defect       1  
4  fixed defect       1

#### Описание признаков

**age**: The person's age in years

**sex**: The person's sex (1 = male, 0 = female)

**cp**: The chest pain experienced (Value 1: typical angina, Value 2: atypical angina, Value 3: non-anginal pain, Value 4: asymptomatic)

**trestbps**: The person's resting blood pressure (mm Hg on admission to the hospital)

**chol**: The person's cholesterol measurement in mg/dl

**fbs**: The person's fasting blood sugar (> 120 mg/dl, 1 = true; 0 = false)

**restecg**: Resting electrocardiographic measurement (0 = normal, 1 = having ST-T wave abnormality, 2 = showing probable or definite left ventricular hypertrophy by Estes' criteria)

**thalach**: The person's maximum heart rate achieved

**exang**: Exercise induced angina (1 = yes; 0 = no)

**oldpeak**: ST depression induced by exercise relative to rest ('ST' relates to positions on the ECG plot. See more here)

**slope**: the slope of the peak exercise ST segment (Value 1: upsloping, Value 2: flat, Value 3: downsloping)

**ca**: The number of major vessels (0-3)

**thal**: A blood disorder called thalassemia (3 = normal; 6 = fixed defect; 7 = reversable defect)

**target**: Heart disease (0 = no, 1 = yes)

> 1. age 
> 2. sex 
> 3. chest pain type (4 values) 
> 4. resting blood pressure 
> 5. serum cholestoral in mg/dl 
> 6. fasting blood sugar > 120 mg/dl
> 7. resting electrocardiographic results (values 0,1,2)
> 8. maximum heart rate achieved 
> 9. exercise induced angina 
> 10. oldpeak = ST depression induced by exercise relative to rest 
> 11. the slope of the peak exercise ST segment 
> 12. number of major vessels (0-3) colored by flourosopy 
> 13. thal: 3 = normal; 6 = fixed defect; 7 = reversable defect

___

**Сколько мужчин в датасете? Сколько женщен? (sex) **

In [4]:
sex_stat = df.groupby('sex').count().target

In [5]:
for s, n in zip(['male', 'female'], ['мужчин', 'женщин']):
    print(n, sex_stat[s])

мужчин 207
женщин 96


 **Какой процент мужчин в датасете? (решите в одну строчку, не используя предыдущий результат. Не забудте знак процента) **

In [6]:
male_percents = df[df['sex'] == 'male'].shape[0]/df.shape[0] * 100

In [29]:
print(f'Мужчины составляют {male_percents:5.2f}% выборки.')

Мужчины составляют 68.32% выборки.


**Сколько мужчин имеют заболевание сердца? Сколько женщин имеют заболевание сердца? **

In [30]:
ilness_stat = df[df['target'] == 1].groupby('sex').count().target

In [32]:
print("Выборка пациентов с заболеванием сердца включает:")
for s, n in zip(['male', 'female'], ['мужчин', 'женщин']):
    print(n, ilness_stat[s])

Выборка пациентов с заболеванием сердца включает:
мужчин 93
женщин 72


**Какую долю, от общего числа пациентов, занимают мужчины не имеющие болезнь сердца?**

In [34]:
healthy_male_percents = df[(df['sex'] == 'male') & (df['target'] == 0)].shape[0]/df.shape[0] * 100

In [36]:
print(f'Мужчины без болезни сердца составляют {healthy_male_percents:5.2f}% выборки.')

Мужчины без болезни сердца составляют 37.62% выборки.


**Сколько лет самому молодому пациенту, страдающему болезнью сердца?**

In [47]:
youngest = df[(df['sex'] == 'male') & (df['target'] == 1)]['age'].min()

In [48]:
print(youngest)

29


**Сколько лет самому возастному пациенту, у которого нет проблем с сердцем?**

In [49]:
healthy_oldest = df[df['target'] == 0]['age'].max()

In [50]:
print("Если пол пациента не берется во внимание, то", healthy_oldest)

Если пол пациента не берется во внимание, то 77


**Сколько лет самой молодой женщине, которая страдает болезнью сердца?**

In [51]:
youngest_ill_lady = df[(df['sex'] == 'female') & (df['target'] == 1)]['age'].min()

In [52]:
print(youngest_ill_lady)

34


**Какой средний возраст женщин?**

In [54]:
ladies_mean_age = df[(df['sex'] == 'female')]['age'].mean()

In [60]:
print(f"{ladies_mean_age:3.1f}")

55.7


**Каковы средние значения и среднеквадратичные отклонения возраста тех, кто страдают болезнью сердца?**

In [117]:
df_ill = df[df['target'] == 1]['age']

In [132]:
print(f"Средний возраст пациентов с болезнью сердца: {df_ill.mean():5.2f}" )
print(f"Среднеквадратичное отклонение возраста пациентов с болезнью сердца: {df_ill.std():5.2f}" )

print("Another way, by using.agg()")
df_groupped_this = df.groupby('target').agg({'age': ['mean', 'std']})
print(f"Средний возраст пациентов с болезнью сердца: {df_groupped_this.loc[1][0]:5.2f}" )
print(f"Среднеквадратичное отклонение возраста пациентов с болезнью сердца: {df_groupped_this.loc[1][1]:5.2f}" )


Средний возраст пациентов с болезнью сердца: 52.50
Среднеквадратичное отклонение возраста пациентов с болезнью сердца:  9.55
Another way, by using.agg()
Средний возраст пациентов с болезнью сердца: 52.50
Среднеквадратичное отклонение возраста пациентов с болезнью сердца:  9.55


** Правда ли, что люди не болеющие болезнью сердца имеют уровня холестерина меньше среднего? (chol) **

In [114]:

answer = df[df['target'] == 0]['cholesterol'].mean() < df['cholesterol'].mean()

In [115]:
print('Правда' if answer else 'Это не так')

Это не так


**Выведите статистику rest_ecg для все числовых признаков, его максимально и среднее значение (используйте groupby(), решите в одну строчку)**

In [32]:
names = [col for col in df.columns if df[col].dtypes in ['int64', 'float64'] and col != 'target']
agg_dict = {}
for n in names:
    agg_dict[n] = ['max', 'mean']
names.append('rest_ecg')
df_this = df_this.loc[:, names].copy(deep=True)    

In [33]:
df_this.groupby('rest_ecg').agg(agg_dict)     

age            resting_blood_pressure  \
                             max       mean                    max   
rest_ecg                                                             
ST-T wave abnormality         71  52.914474                    180   
left ventricular hypertrophy  76  61.000000                    180   
normal                        77  55.687075                    200   

                                         cholesterol              \
                                    mean         max        mean   
rest_ecg                                                           
ST-T wave abnormality         129.065789         354  237.269737   
left ventricular hypertrophy  140.500000         327  261.750000   
normal                        134.027211         564  255.142857   

                             max_heart_rate_achieved              \
                                                 max        mean   
rest_ecg                                                           
ST-T wave abnormality                            194  151.960526   
left ventricular hypertrophy                     140  125.750000   
normal                                           202  147.904762   

                             st_depression           num_major_vessels  \
                                       max      mean               max   
rest_ecg                                                                 
ST-T wave abnormality                  5.6  0.879605                 4   
left ventricular hypertrophy           4.4  2.725000                 3   
normal                                 6.2  1.159184                 4   

                                        
                                  mean  
rest_ecg                                
ST-T wave abnormality         0.638158  
left ventricular hypertrophy  1.000000  
normal                        0.816327

**Посчитайте у кого уровень депрессии при физический нагрузке выше (в среднем), среди мужчин страдающих болезнью сердца или среди женщин не страдающих болезнью сердца (st_depression)**

In [36]:
result = df[(df['sex'] == 'male') & (df['target'] == 1)]['st_depression'].mean() < df[(df['sex'] == 'female') & (df['target'] == 0)]['st_depression'].mean()

In [37]:
print(f'В среднем уровень депрессии выше у {"женщин, не страдающих болезнью сердца" if result else "мужчин, страдающих болезнью сердца"}')

В среднем уровень депрессии выше у женщин, не страдающих болезнью сердца


**Посчтитайте максимальный и минимальный уровень холестерина для каждого типа chest_pain_type, rest_ecg, thalassemia. Пишите код оптимально, можно использовать циклы**

In [47]:
df.groupby(['chest_pain_type', 'rest_ecg', 'thalassemia']).agg({'cholesterol': ['max', 'min']})

cholesterol  \
                                                                        max   
chest_pain_type  rest_ecg                     thalassemia                     
atypical angina  ST-T wave abnormality        fixed defect              340   
                                              normal                    218   
                                              reversable defect         313   
                 left ventricular hypertrophy fixed defect              197   
                 normal                       fixed defect              417   
                                              normal                    256   
                                              reversable defect         564   
non-anginal pain ST-T wave abnormality        fixed defect              240   
                                              reversable defect         298   
                 normal                       fixed defect              283   
                                              normal                    233   
                                              reversable defect         288   
typical angina   ST-T wave abnormality        fixed defect              354   
                                              normal                    315   
                                              reversable defect         353   
                 left ventricular hypertrophy fixed defect              327   
                                              normal                    318   
                                              reversable defect         205   
                 normal                       fixed defect              394   
                                              normal                    276   
                                              reversable defect         409   

                                                                      
                                                                 min  
chest_pain_type  rest_ecg                     thalassemia             
atypical angina  ST-T wave abnormality        fixed defect       141  
                                              normal             218  
                                              reversable defect  126  
                 left ventricular hypertrophy fixed defect       197  
                 normal                       fixed defect       149  
                                              normal             216  
                                              reversable defect  212  
non-anginal pain ST-T wave abnormality        fixed defect       204  
                                              reversable defect  199  
                 normal                       fixed defect       182  
                                              normal             186  
                                              reversable defect  193  
typical angina   ST-T wave abnormality        fixed defect       149  
                                              normal             169  
                                              reversable defect  131  
                 left ventricular hypertrophy fixed defect       327  
                                              normal             318  
                                              reversable defect  205  
                 normal                       fixed defect       166  
                                              normal             176  
                                              reversable defect  164

In [62]:
param = 'cholesterol'
for keyword in ['chest_pain_type', 'rest_ecg','thalassemia']: 
    print(keyword)
    for kind in set(df[keyword].tolist()):
        data_this = df[df[keyword] == kind][param] 
        print(f"\t{kind} {param} level: \n\t\t min = {data_this.min()}\n\t\t max = {data_this.max()}" )

chest_pain_type
	atypical angina cholesterol level: 
		 min = 126
		 max = 564
	non-anginal pain cholesterol level: 
		 min = 182
		 max = 298
	typical angina cholesterol level: 
		 min = 131
		 max = 409
rest_ecg
	left ventricular hypertrophy cholesterol level: 
		 min = 197
		 max = 327
	ST-T wave abnormality cholesterol level: 
		 min = 126
		 max = 354
	normal cholesterol level: 
		 min = 149
		 max = 564
thalassemia
	reversable defect cholesterol level: 
		 min = 126
		 max = 564
	fixed defect cholesterol level: 
		 min = 141
		 max = 417
	normal cholesterol level: 
		 min = 169
		 max = 318


**Сколько значений может принимать каждый из категориальных признаков?**

In [75]:
for keyword in ['chest_pain_type', 'rest_ecg','thalassemia']:
    all_values = set(df[keyword].tolist())
    print(f"{keyword} takes {len(all_values)} values: {all_values}")

chest_pain_type takes 3 values: {'atypical angina', 'non-anginal pain', 'typical angina'}
rest_ecg takes 3 values: {'left ventricular hypertrophy', 'ST-T wave abnormality', 'normal'}
thalassemia takes 3 values: {'reversable defect', 'fixed defect', 'normal'}


In [68]:
print("см. выше")

см. выше


**У какого категориального признака наблюдается самый сильный дизбаланс классов?**

In [112]:
cols = ['chest_pain_type', 'rest_ecg', 'thalassemia', 'target'] 
df_example = df.loc[:, cols]
max_disbalanced = None
for keyword in cols[:-1]: 
    print(keyword)  
    df_groupped = df_example.groupby(keyword).count() 
    dis = (df_groupped['target'].max() - df_groupped['target'].min()) / df_groupped['target'].sum()
    if not max_disbalanced:
        max_disbalanced = {"property": keyword, "value": dis}
    else:
        if dis > max_disbalanced.get("value"):
            max_disbalanced = {"property": keyword, "value": dis}
    print(f"количество записей в классе\n\t наименьшее {df_groupped['target'].min()}, наибольшее {df_groupped['target'].max()}" )


chest_pain_type
количество записей в классе
	 наименьшее 23, наибольшее 193
rest_ecg
количество записей в классе
	 наименьшее 4, наибольшее 152
thalassemia
количество записей в классе
	 наименьшее 20, наибольшее 166


In [113]:
print(f"\nПризнак с самым сильным диCбалансом классов: {max_disbalanced.get('property')}")    


Признак с самым сильным диCбалансом классов: chest_pain_type


---